### PR TITLE
[SDK-571] Fixes percent encoding with phdp query expression

### PR DESCRIPTION
**Pull request by alessioborraccino**
## Motivation and Context

see issue here: [SDK-570](https://gesundheitscloud.atlassian.net/browse/SDK-570)

## How is it being implemented?

Removes the double api call on search and count because not needed anymore, as one call is now enough. In order to specify the expression query, the app also introduces a special struct. `TagsParameter`.
It represents the "tags" parameter in search/count and upload api calls queries, due to the fact that it is mapped differently according to different cases.
The `TagGroup` is able to create `TagsParameter`s according to which api operation is being called and if legacy tags need to be supported.  

Side effects done with this PR:
- Fetch methods (like count before) can also now be called without specifying the type with the same result
- Introduces another SDKError: invalidCharacterInTag, as addingPercentEncoding might fail in some edge cases (namely some malformed UTF8 cluster of characters)

## How Has This Been Tested?

Adds new tests for all the different cases of parameters, adjusted all failing cases due to different paths, ran and basic tested on example app.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have updated the changelog accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: SDK/Sources/Services/RecordService/TagsParameter.swift [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    enum OperationType {
+        case search(supportingLegacyTags: Bool = true)
+        case upload
+    }
+
+    struct OrComponent: Hashable {
+        let formattedTag: String
+
+        init(key: String, value:String, separator: Character = "=") {
+            self.formattedTag = "\(key)\(separator)\(value)"
```
[the remark this answers is not in the document]
Comment (alessioborraccino) [reply SDK/Sources/Services/RecordService/TagsParameter.swift:29]: `keySeparatorValue`.
` \(something)` is the syntax for string interpolation in swift

[the remark this answers is not in the document]
Comment (bitPogo) [reply SDK/Sources/Services/RecordService/TagsParameter.swift:29]: thx

**Review by ztepsa — APPROVED** [review]
Seems good 🚀

File: SDK/Sources/Services/RecordService/RecordService.swift [hunk 2]
```diff
@@ -350,11 +289,22 @@ extension RecordService {
...
+                parameters["tags"] = encryptedTagsParameters.asTagExpressions.joined(separator: ",")
             }
-
             return parameters
         }
     }
 }
+
+// MARK: - Helpers
+extension CryptoServiceType {
```
[the remark this answers is not in the document]
Comment (alessioborraccino) [reply SDK/Sources/Services/RecordService/RecordService.swift:303]: apparently it does

File: SDK/Tests/ModelTests/TagsParameterTests.swift [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+fileprivate extension TagsParameter {
+    static var stringInitialized: TagsParameter = TagsParameter("tag=value")
+    static var keyValueInitialized: TagsParameter = TagsParameter(TagsParameter.OrComponent(key: "tag", value: "value", separator: "="))
+    static var doubleComponentInitialized: TagsParameter = TagsParameter([TagsParameter.OrComponent(formattedTag: "tag=value"), TagsParameter.OrComponent(formattedTag: "tag2=value2")])
+}
+
+class TagsParameterTests: XCTestCase {
+
+    func testConvenienceInitaliazers() throws {
```
[the remark this answers is not in the document]
Comment (alessioborraccino) [reply SDK/Tests/ModelTests/TagsParameterTests.swift:27]: fixed

**Review by jose-cuadros-data4life — APPROVED** [review]
Nice 👍
